GLTFExporter: Export images as binary in GLB.

donmccurdy opened the pull request:
~~Work in progress, I'd like to (1) get feedback and (2) test this a bit more.~~

The main change here is that when `options.binary === true` and `options.embedImages === true`, the images will be written to bufferViews instead of Data URIs, taking up less space. For the example scene, it's a ~25% file size difference.

I wasn't sure how to manage the async parts, since the processing so far hasn't needed as much of that. Didn't want to add callbacks to everything, and wasn't sure it was worth it to add Promises just for this change, but I could imagine the `next()` counters I'm using here might be confusing to read... Any preferences?

/cc @fernandojsg

Commit:
GLTFExporter: WIP allow images to be stored in bufferViews.

Commit:
GLTFExporter: Fix bufferview image async handling, encoding.

takahirox commented:
>  how to manage the async parts

Voting for Promises.

fernandojsg commented:
I'd vote for promises too. I'd also add an option to define the behaviour for the images if the user wants to export then as data URIs or bufferView.
Btw this changes makes obsolete the proposal and changes on https://github.com/mrdoob/three.js/pull/13419 but I guess it will take more time to review :)

Commit:
GLTFExporter: Revert to dev.

Commit:
GLTFExporter: Re-apply image buffer changes.

donmccurdy commented:
Rebased and switched to Promises, thanks!

I don't think we want an option for data URIs instead of bufferView; they use ~30% more space in a GLB with no benefit I'm aware of.

takahirox commented on examples/js/exporters/GLTFExporter.js:
```diff
@@ -366,6 +347,45 @@ THREE.GLTFExporter.prototype = {
+		 * @param {Blob} blob
+		 * @return {Promise<Integer>}
+		 */
+		function processBufferViewImage ( blob ) {
```
You don't need an space after function name

Commit:
GLTFExporter: Whitespace cleanup.

fernandojsg commented:
@donmccurdy regading the data uri option I was thinking on engines that doesn't have implemented bufferview for images, but I'm not aware of the current state of most of the engines so you may know more about it.

donmccurdy commented:
It is part of the core spec, engines should consider it a requirement and I'm not aware of any that don't support this. Inclined to avoid the extra API surface unless the issue is brought up.

fernandojsg commented:
Yep, so in that case 👎 to the extra params

mrdoob commented:
Is this ready to go then?

takahirox commented on examples/js/exporters/GLTFExporter.js:
```diff
@@ -366,6 +347,45 @@ THREE.GLTFExporter.prototype = {
+
+			}
+
+			return new Promise( function (resolve) {
```
Need spaces before and after `resolve`.

donmccurdy replied:
Done ✅

takahirox commented on examples/js/exporters/GLTFExporter.js:
```diff
@@ -366,6 +347,45 @@ THREE.GLTFExporter.prototype = {
+
+				var reader = new window.FileReader();
+				reader.readAsArrayBuffer( blob );
+				reader.addEventListener('loadend', function () {
```
Needs a space before `'loadend'`.

takahirox replied:
And how about using `.onloadend` for the consistency with other parts in the exporter?

donmccurdy replied:
Done ✅

donmccurdy committed:
GLTFExporter: Whitespace cleanup.

fernandojsg commented:
@donmccurdy looking great! 👏

mrdoob commented:
Sweet! Thanks!
